### ISSUE TITLE
Intelligent photo & video search

### ISSUE THREAD

**Issue by mholt** [timelinize/timelinize#7]
Integrate a model for a photo/video search that can identify objects or setting/description of photos and videos for smarter search. Of course, it needs to run on-device.

**Comment by mholt** [comment]
I have gotten a state-of-the-art model from Hugging Face up and running, and from a few cursory tests, it looks very promising.

I even tried it with a picture of an adhesive label that says "The radio" ... and for various search terms got these scores:

- 0.9822: `a sticker that says "the radio"`
- 0.017: `adhesive label`
- 0.135: `sticker`
- 0.0125: `the radio`
- 0.0: `desk`, `boat`, `foliage`, `beautiful sunset`

This is better than I had hoped for!

I will need to figure out how to make this work in a photo search scenario... I suspect I can extract the embeddings, put them in the database, and perform similarity searches using sqlite-vec: https://github.com/asg017/sqlite-vec (more info: https://alexgarcia.xyz/blog/2024/building-new-vector-search-sqlite/index.html)

Also need to explore video searching... and figure out how to make this part of the distribution:

- I installed [`uv`](https://github.com/astral-sh/uv) (easy)
- I created a uv project and added dependencies/tools like pytorch and Pillow (a bit trickier, but now I think I know how)
- Ran a 15-line python script that does the classification (wow!)

I will likely need to run a Flask server in a Python program, and have it run alongside Timelinize, so that I can use HTTP socket to quickly run Python code concurrently, rather than spinning up `uv run ...` for every image that comes in through the processor.

**Comment by mholt** [comment]
The above tests were with the classifier, but I haven't been able to get such good results with a more practical "search" approach, where image embeddings are stored in the DB, then search queries (either text or image; because SigLIP is multi-modal) are evaluated to get their embedding, then we find the images with the closest embeddings...

Anyway, that approach isn't giving me quite as good results for some reason, I have some intuition as to why, but looking into it more.

This person writes up very good results, so I'm probably just doing it wrong: https://harper.blog/2024/04/12/i-accidentally-built-a-meme-search-engine/

**Comment by mholt** [comment]
I have good search results now:

- I had to use a concrete model class instead of `AutoModel`
- I also got sqlite-vec working, though that is a WIP project itself, so it may have bugs or not perform super fast yet. Still, I think long-term, this is a good bet.

My main qualm with sqlite-vec is that the "distance" values are arbitrary numbers, whereas with my Python code, I was able to get it to give me similarity scores from 0-1, where 0 is "no match" and 1 is "identical", and the scores cumulatively add up to 1. This makes for very intuitive search results. I don't know if sqlite-vec does this, but I [opened an issue](https://github.com/asg017/sqlite-vec/issues/115) to ask. It's not _strictly_ required, but will make it much easier to filter non-results, and to show intuitive confidence scores to the user.

The nice thing about this approach is that search queries can be text or images. So this adds a lot of capability to Timelinize.

I still need to figure out how to bundle Python with Timelinize, but then it should be fairly straightforward. Man. These ML models have gotten WAY better since I was in grad school studying this stuff.

This will also pave the way for #6 -- face recognition, where we can automatically associate people found in your photos with your timeline.

**Comment by mholt** [comment]
I have implemented a rudimentary, but helpful-enough, photo search using the state-of-the-art Siglip model and a sidecar Python environment: https://github.com/timelinize/timelinize/commit/96d967eb5afeb3fa1fa86cf59aff7294661bc07d

It will auto-install [`uv`](https://github.com/astral-sh/uv) and all the other dependencies, but Timelinize will have to be restarted in order for it to actually run the ML server. I am still trying to figure out a way to avoid that.

Anyway, if you go to the Gallery page, you have a new "Search" text box where you can type a query. Give it 5-15 seconds.

![image](https://github.com/user-attachments/assets/8a26f42a-ac0b-40f9-8ddc-447b33f9230b)

Seems to be pretty good with some semantics, not so good at others. We have a lot of fine-tuning and optimizing to do! But it's a good start. This same logic can be applied to text items, as well as search-by-image features.

Still need to figure out video support. This may require separate models, not sure yet.

If you do try out this branch (currently in `new-imports`), please be patient with it, and feel free to contribute code fixes and feedback - thanks!

**Comment by mholt** [comment]
I've since reduced the model size so it runs smoother on more commodity hardware.
